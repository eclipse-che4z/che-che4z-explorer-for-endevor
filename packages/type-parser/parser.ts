/*
 * Â© 2021 Broadcom Inc and/or its subsidiaries; All rights reserved
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Broadcom, Inc. - initial API and implementation
 */

import * as t from 'io-ts';
import { PathReporter } from 'io-ts/PathReporter';
import { fold } from 'fp-ts/lib/Either';

/**
 * Verifies that `input` is a valid `type`
 * @param type
 * @param input
 */
export const parseToType = <T, O, I>(type: t.Type<T, O, I>, input: I): T => {
  const result = type.decode(input);
  return fold<t.Errors, T, T>(
    (_errors) => {
      // validation failed, tell the user why `input` does not match `type`
      const messages = PathReporter.report(result);
      throw new Error(messages.join('\n'));
    },
    (value) => value
  )(result);
};
