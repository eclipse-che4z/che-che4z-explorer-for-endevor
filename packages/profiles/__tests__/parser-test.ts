/*
 * Â© 2023 Broadcom Inc and/or its subsidiaries; All rights reserved
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Broadcom, Inc. - initial API and implementation
 */

import { parseToType } from '@local/type-parser/parser';
import {
  BaseProfile,
  EndevorLocationProfile,
  EndevorServiceProfile,
} from '../_ext/Profile';

describe('parseToType profiles', () => {
  // here can be any profile: Base, EndevorService or Location
  describe('common profile cases', () => {
    it('should not report for missed fields', () => {
      // arrange
      const emptyProfile: BaseProfile = {};
      // act
      const actualBaseProfile = parseToType(BaseProfile, emptyProfile);
      // assert
      expect(actualBaseProfile).toEqual(emptyProfile);
    });
    it('should report for wrong typed profile', () => {
      // arrange
      const wrongTypedProfile = {
        host: 42,
      };
      // act && assert
      expect(() =>
        parseToType(BaseProfile, wrongTypedProfile)
      ).toThrowErrorMatchingSnapshot();
    });
  });
  describe('parseToType base profile', () => {
    it('should parse proper base profile', () => {
      // arrange
      const baseProfile: BaseProfile = {
        host: 'host',
        port: 9090,
        user: 'some_user',
        password: 'some_pass',
      };
      // act
      const actualBaseProfile = parseToType(BaseProfile, baseProfile);
      // assert
      expect(actualBaseProfile).toEqual(baseProfile);
    });
  });
  describe('parseToType endevor service profile', () => {
    it('should parse proper endevor service profile', () => {
      // arrange
      const endevorServiceProfile: EndevorServiceProfile = {
        host: 'host',
        port: 9090,
        user: 'some_user',
        password: 'some_pass',
      };
      // act
      const actualProfile = parseToType(
        EndevorServiceProfile,
        endevorServiceProfile
      );
      // assert
      expect(actualProfile).toEqual(endevorServiceProfile);
    });
  });
  describe('parseToType endevor location profile', () => {
    it('should parse proper endevor location profile', () => {
      // arrange
      const endevorLocationProfile: EndevorLocationProfile = {
        instance: 'NDVRCNFG',
        system: 'SYS',
      };
      // act
      const actualProfile = parseToType(
        EndevorLocationProfile,
        endevorLocationProfile
      );
      // assert
      expect(actualProfile).toEqual(endevorLocationProfile);
    });
  });
});
