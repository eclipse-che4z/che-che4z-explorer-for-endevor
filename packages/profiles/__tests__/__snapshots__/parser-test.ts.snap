// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseToType profiles common profile cases should report for wrong typed profile 1`] = `"Invalid value 42 supplied to : Partial<{ host: string, port: number, user: string, password: string, tokenValue: string, tokenType: TokenType }>/host: string"`;
